EIPS/eip-2200.md

```diff
@@ -0,0 +1,322 @@
+
+The original definition of EIP-1283 created a danger of a new kind of
+reentrancy attacks on existing contracts as Solidity by default grants
+a "stipend" of 2300 gas to simple transfer calls. This danger is
+easily mitigated if `SSTORE` is not allowed in low gasleft state,
+without breaking the backward compatibility and the original intention
+of EIP-1283.
+
+This EIP also replaces the original EIP-1283 value definitions of gas
+by parameters, so that it's more structured, and easier to define
+changes in the future.
+
+## Specification
+
+Define variables `SLOAD_GAS`, `SSTORE_SET_GAS`, `SSTORE_RESET_GAS` and
+`SSTORE_CLEARS_SCHEDULE`. The old and new values for those variables
+are:
+
+* `SLOAD_GAS`: changed from `200` to `800`.
+* `SSTORE_SET_GAS`: `20000`, not changed.
+* `SSTORE_RESET_GAS`: `5000`, not changed.
+* `SSTORE_CLEARS_SCHEDULE`: `15000`, not changed.
+
+Change the definition of EIP-1283 using those variables. The new
+specification, combining EIP-1283 and EIP-1706, will look like
+below. The terms *original value*, *current value* and *new value* are
+defined in EIP-1283.
+
+Replace `SSTORE` opcode gas cost calculation (including refunds) with
+the following logic:
+
+* If *current value* equals *new value* (this is a no-op), `SLOAD_GAS`
+  is deducted.
+* If *current value* does not equal *new value*
+  * If *original value* equals *current value* (this storage slot has
+    not been changed by the current execution context)
+    * If *original value* is 0, `SSTORE_SET_GAS` is deducted.
+    * Otherwise, `SSTORE_RESET_GAS` gas is deducted. If *new value* is
+      0, add `SSTORE_CLEARS_SCHEDULE` gas to refund counter.
+  * If *original value* does not equal *current value* (this storage
+    slot is dirty), `SLOAD_GAS` gas is deducted. Apply both of the
+    following clauses.
+    * If *original value* is not 0
+      * If *current value* is 0 (also means that *new value* is not
+        0), remove `SSTORE_CLEARS_SCHEDULE` gas from refund
+        counter. We can prove that refund counter will never go below
+        0.
+      * If *new value* is 0 (also means that *current value* is not
+        0), add `SSTORE_CLEARS_SCHEDULE` gas to refund counter.
+    * If *original value* equals *new value* (this storage slot is
+      reset)
+      * If *original value* is 0, add `SSTORE_SET_GAS - SLOAD_GAS` to
+        refund counter.
+      * Otherwise, add `SSTORE_RESET_GAS - SLOAD_GAS` gas to refund
+        counter.
+* If *gasleft* is less than or equal to gas stipend, fail the current
+  call frame with 'out of gas' exception.
+
+An implementation should also note that with the above definition, if
+the implementation uses call-frame refund counter, the counter can go
+negative. If the implementation uses transaction-wise refund counter,
+the counter always stays positive.
+
+## Rationale
+
+This EIP mostly achieves what a transient storage tries to do
+(EIP-1087 and EIP-1153), but without the complexity of introducing the
```

Comment:
Should make these two into link to the EIP.

Reply:
Fixed.